onnxruntime/core/providers/cpu/nn/tfidfvectorizer.cc

```diff
@@ -25,168 +27,57 @@ ONNX_CPU_OPERATOR_KERNEL(
-      RunningHash(*first);
-      items_.push_back(*first);
-      ++first;
-    }
-    assert(!items_.empty());
-  }
-  // For sampling
-  explicit NgramEntry() : NgramEntryBase(0) {}
-  void AddItem(int64_t v) {
-    items_.push_back(v);
-    RunningHash(v);
-  }
-  void DebugPrint() const {
-    std::copy(items_.cbegin(), items_.cend(), std::ostream_iterator<int64_t>(std::cout, ","));
-    std::cout << std::endl;
-  }
-  void Clear() {
-    items_.clear();
-    hash_ = 0;
-  }
-  bool operator==(const NgramEntry& o) const {
-    return items_ == o.items_;
-  }
-  size_t Hash() const {
-    return hash_;
-  }
-};
+using StrMap = std::unordered_map<std::reference_wrapper<const std::string>, NgramPart<std::string>,
+                                  std::hash<std::string>, std::equal_to<std::string>>;
 
 template <>
-class NgramEntry<int32_t> : public NgramEntry<int64_t> {
- public:
-  template <typename ForwardIter>
-  explicit NgramEntry(size_t id, ForwardIter first, ForwardIter last) : NgramEntry<int64_t>(id, first, last) {}
-  explicit NgramEntry() = default;
+struct NgramPart<std::string> {
+  size_t id_;  // 0 - means no entry, search for a bigger N
+  StrMap leafs_;
+  explicit NgramPart(size_t id) : id_(id) {}
 };
 
-template <>
-class NgramEntry<std::string> : public NgramEntryBase {
- private:
-  std::vector<std::reference_wrapper<const std::string>> items_;
-  size_t hash_ = 0;
-
-  void RunningHash(const std::string& s) {
-    std::hash<std::string> hf;
-    hash_ ^= hf(s) + 0x9e3779b9 + (hash_ << 6) + (hash_ >> 2);
-  }
-
- public:
-  template <typename ForwardIter>
-  explicit NgramEntry(size_t id, ForwardIter first, ForwardIter last) : NgramEntryBase(id) {
-    while (first != last) {
-      RunningHash(*first);
-      items_.push_back(std::cref(*first));
+// Returns next ngram_id
+template <class ForwardIter, class Map>
+inline size_t PopulateGrams(ForwardIter first, size_t ngrams, size_t ngram_size, size_t ngram_id, Map& c) {
+  for (; ngrams > 0; --ngrams) {
+    size_t n = 1;
+    Map* m = &c;
+    while (true) {
+      auto p = m->emplace(*first, 0);
       ++first;
-    }
-    assert(!items_.empty());
-  }
-  explicit NgramEntry() : NgramEntryBase(0) {}
-  void AddItem(const std::string& s) {
-    items_.push_back(std::cref(s));
-    RunningHash(s);
-  }
-  void DebugPrint() const {
-    std::copy(items_.cbegin(), items_.cend(), std::ostream_iterator<std::string>(std::cout, ","));
-    std::cout << std::endl;
-  }
-  void Clear() {
-    items_.clear();
-    hash_ = 0;
-  }
-
-  bool operator==(const NgramEntry& o) const {
-    if (items_.size() == o.items_.size()) {
-      std::equal_to<std::string> pred;
-      for (size_t i = 0; i < items_.size(); ++i) {
-        if (!pred(items_[i], o.items_[i])) {
-          return false;
-        }
+      if (n == ngram_size) {
+        ORT_ENFORCE(p.first->second.id_ == 0, "Duplicate ngram detected, size: ", ngram_size, " id: ", ngram_id);
+        p.first->second.id_ = ngram_id;
```

Comment:
Would it be simpler to just check that p.second is true given that returns whether it was inserted or not?

Reply:
The way it is structured is like a Trie.  For example, if you have a 3-gram (1,2,3) and Unigram (1) existing but not a Bi-gram (1,2). So one would insert (1).  Then it would insert (1,2,3) with a valid id. 
If someone come in and inserts (1,2,4) then node for 2 would already exist but it is 0 bc 2-gram (1,2) does not exist.
It really speeds up even unit tests visibly although they are not complex by any means.

---
In reply to: [391957470](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r391957470) [](ancestors = 391957470)